mycpp/mark_sweep_heap.h

```diff
@@ -62,6 +65,134 @@ class MarkSet {
 };
 
+#ifdef POOL_ALLOC
+// A simple Pool allocator for allocating small objects. It maintains an ever
+// growing number of Blocks each consisting of a number of fixed size Cells.
+// Memory is handed out one Cell at a time.
+// Note: within the context of the Pool allocator we refer to object IDs as cell
+// IDs because in addition to identifying an object they're also used to index
+// into the Cell storage.
+template <int CellsPerBlock, size_t CellSize>
+class Pool {
+ public:
+  static constexpr size_t kMaxObjSize = CellSize;
+
+  Pool() = default;
+
+  void* Allocate(int* obj_id) {
+    num_allocated_++;
+
+    if (!free_list_) {
+      // Allocate a new Block and add every new Cell to the free list.
+      Block* block = static_cast<Block*>(malloc(sizeof(Block)));
+      blocks_.push_back(block);
+      bytes_allocated_ += kBlockSize;
+      num_free_ += CellsPerBlock;
+
+      // The starting cell_id for Cells in this block.
+      int cell_id = (blocks_.size() - 1) * CellsPerBlock;
+      for (Cell& cell : block->cells) {
+        FreeCell* free_cell = reinterpret_cast<FreeCell*>(cell.data());
+        free_cell->id = cell_id++;
+        free_cell->next = free_list_;
+        free_list_ = free_cell;
+      }
+    }
+
+    FreeCell* cell = free_list_;
+    free_list_ = free_list_->next;
+    num_free_--;
+    *obj_id = cell->id;
+    return cell;
+  }
+
+  void PrepareForGc() {
+    DCHECK(!gc_underway_);
+    gc_underway_ = true;
+    mark_set_.ReInit(blocks_.size() * CellsPerBlock);
+  }
+
+  bool IsMarked(int cell_id) {
+    DCHECK(gc_underway_);
+    return mark_set_.IsMarked(cell_id);
+  }
+
+  void Mark(int cell_id) {
+    DCHECK(gc_underway_);
+    mark_set_.Mark(cell_id);
+  }
+
+  void Sweep() {
+    DCHECK(gc_underway_);
+    // Iterate over every Cell linking the free ones into a new free list.
+    num_free_ = 0;
+    free_list_ = nullptr;
+    int cell_id = 0;
+    for (Block* block : blocks_) {
+      for (Cell& cell : block->cells) {
+        if (!mark_set_.IsMarked(cell_id)) {
+          num_free_++;
+          FreeCell* free_cell = reinterpret_cast<FreeCell*>(cell.data());
+          free_cell->id = cell_id;
+          free_cell->next = free_list_;
+          free_list_ = free_cell;
+        }
+        cell_id++;
+      }
+    }
+    gc_underway_ = false;
+  }
+
+  void Free() {
+    for (Block* block : blocks_) {
+      free(block);
+    }
+    blocks_.clear();
+  }
+
+  int num_allocated() {
+    return num_allocated_;
+  }
+
+  int64_t bytes_allocated() {
+    return bytes_allocated_;
+  }
+
+  int num_live() {
+    return blocks_.size() * CellsPerBlock - num_free_;
+  }
+
+ private:
+  using Cell = std::array<uint8_t, CellSize>;
+
+  struct Block {
+    std::array<Cell, CellsPerBlock> cells;
+  };
+
+  // Unused/free cells are tracked via a linked list of FreeCells. The FreeCells
+  // are stored in the unused Cells, so it takes no extra memory to track them.
```

Comment:
OK this is pretty interesting

One thing I thought of was using the complement of the MarkSet as the free list.  As long as mutating/allocating and mark/sweep are distinct phases, and they are, that should work.



But I think this is probably even better!

Reply:
Oh right, and on allocation you do a linear scan of the mark set? I considered that, worth a try for sure. Might be faster